### DISCUSSION TITLE
rt-detr with custom dataset training cannot convert to int8.

### DISCUSSION THREAD

**Discussion by harufumigithub** [ultralytics/ultralytics#12698] [Q&A]
**I have trained a custom dataset with following:**
model = RTDETR('yolov8n-rtdetr.yaml')
model.load('yolov8n.pt')
results = model.train(data="./tag_detection.yaml", device=0, imgsz=640, epochs=1000, batch=8)   

**After successfully trained and running inference works.  I tried to convert best.pt to tflite as follows:**
'yolo export model=best.pt format=tflite int8=True' 

**However conversion goes successful up to generating float16.tflite but failed to generate int8.tflite as follows:**
Ultralytics YOLOv8.2.2 🚀 Python-3.9.13 torch-2.2.1 CPU (11th Gen Intel Core(TM) i7-11700B 3.20GHz)
YOLOv8n-rtdetr summary: 304 layers, 9483256 parameters, 0 gradients, 16.7 GFLOPs

PyTorch: starting from 'runs/detect/train/weights/best.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 300, 8) (72.9 MB)

TensorFlow SavedModel: starting export with tensorflow 2.13.1...

ONNX: starting export with onnx 1.15.0 opset 17...
ONNX: simplifying with onnxsim 0.4.36...
ONNX: export success ✅ 5.1s, saved as 'runs/detect/train/weights/best.onnx' (36.6 MB)
TensorFlow SavedModel: starting TFLite export with onnx2tf 1.17.5...

Automatic generation of each OP name started ========================================
Automatic generation of each OP name complete!

Model loaded ========================================================================

Model conversion started ============================================================
WARNING: The optimization process for shape estimation is skipped because it contains OPs that cannot be inferred by the standard onnxruntime.
WARNING: module 'onnx' has no attribute '_serialize'
2024-05-14 22:05:30.677030: E tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:268] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
saved_model output started ==========================================================
saved_model output complete!
Float32 tflite output complete!
Float16 tflite output complete!
Input signature information for quantization
signature_name: serving_default
input_name.0: images shape: (1, 640, 640, 3) dtype: <dtype: 'float32'>
./run.sh: line 12:  8661 Segmentation fault      (core dumped) yolo export model=runs/detect/train/weights/best.pt format=tflite int8=True

**Comment by glenn-jocher** [comment]
@harufumigithub it looks like the int8 quantization process failed because it couldn't utilize a CUDA device, as indicated by the error: **CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected**. This suggests that either the environment you're running the export command in doesn't have access to a CUDA GPU or the setup is not correctly configured to utilize the GPU.

For solving this issue, please ensure the following:
- You have a CUDA-capable GPU installed and it is recognized by your system.
- Your CUDA installation is properly configured.
- The TensorFlow version you are using is compatible with the installed CUDA and cuDNN versions.

If you are running this on a system without a GPU, you will need to switch to a machine with a CUDA-capable GPU to perform int8 quantization, as it is heavily dependent on the GPU for the computation required.

Here's a quick command to check if TensorFlow is detecting your GPU:
```python
import tensorflow as tf
print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))
```

If it returns `Num GPUs Available: 0`, it confirms that TensorFlow isn't detecting your GPU. You might need to troubleshoot your CUDA installation or TensorFlow setup.

**Comment by harufumigithub** [comment]
GPU seems available.  
**$ python3 -c "import tensorflow as tf; print('Num GPUs Available: ', len(tf.config.experimental.list_physical_devices('GPU')))"**
2024-05-15 06:45:42.664954: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-05-15 06:45:42.692092: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-05-15 06:45:43.130783: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
2024-05-15 06:45:43.707841: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-05-15 06:45:43.728940: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-05-15 06:45:43.729066: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
Num GPUs Available:  1

**$ yolo export model=runs/detect/train/weights/best.pt format=tflite int8=true** 
Ultralytics YOLOv8.2.2 🚀 Python-3.9.19 torch-2.2.1 CPU (11th Gen Intel Core(TM) i7-11800H 2.30GHz)
YOLOv8n-rtdetr summary: 304 layers, 9483256 parameters, 0 gradients, 16.7 GFLOPs

PyTorch: starting from 'runs/detect/train/weights/best.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 300, 8) (72.9 MB)

TensorFlow SavedModel: starting export with tensorflow 2.13.1...

ONNX: starting export with onnx 1.16.0 opset 17...
ONNX: simplifying with onnxsim 0.4.36...
ONNX: export success ✅ 4.2s, saved as 'runs/detect/train/weights/best.onnx' (36.6 MB)
TensorFlow SavedModel: starting TFLite export with onnx2tf 1.17.5...

Automatic generation of each OP name started ========================================
Automatic generation of each OP name complete!

Model loaded ========================================================================

Model conversion started ============================================================
WARNING: The optimization process for shape estimation is skipped because it contains OPs that cannot be inferred by the standard onnxruntime.
WARNING: module 'onnx' has no attribute '_serialize'
2024-05-15 06:46:29.507859: E tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:268] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
saved_model output started ==========================================================
saved_model output complete!
Float32 tflite output complete!
Float16 tflite output complete!
Input signature information for quantization
signature_name: serving_default
input_name.0: images shape: (1, 640, 640, 3) dtype: <dtype: 'float32'>
Segmentation fault (core dumped)

**Reply by glenn-jocher** [reply]
It looks like TensorFlow is able to recognize the GPU, but it seems there’s an issue with CUDA during the model export for INT8 quantization, indicating that perhaps TensorFlow wasn’t able to use the GPU during the conversion process.

To troubleshoot, please ensure that:
- CUDA and cuDNN are properly installed and configured.
- Your TensorFlow installation is compatible with the installed CUDA and cuDNN versions.
- The CUDA device is accessible (you might try running the command again or rebooting your system to resolve any transient issues).

If these suggestions don't resolve the issue, provide the TensorFlow, CUDA, and cuDNN version details for further diagnosis. This could help in pinpointing the issue more precisely.

**Comment by harufumigithub** [comment]
Thank you for quick response. Here are my environment. 
$  yolo checks
WARNING ⚠️ Ultralytics settings reset to default values. This may be due to a possible problem with your settings or a recent ultralytics package update. 
View settings with 'yolo settings' or at '/home/trupass/.config/Ultralytics/settings.yaml'
Update settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
Ultralytics YOLOv8.2.2 🚀 Python-3.9.19 torch-2.2.1 CUDA:0 (NVIDIA GeForce RTX 3070 Laptop GPU, 7959MiB)
Setup complete ✅ (16 CPUs, 62.4 GB RAM, 110.8/245.0 GB disk)

OS                  Linux-6.5.0-28-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.9.19
Install             git
RAM                 62.44 GB
CPU                 11th Gen Intel Core(TM) i7-11800H 2.30GHz
CUDA                12.1

matplotlib          ✅ 3.8.4>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.13.0>=1.4.1
torch               ✅ 2.2.1>=1.8.0
torchvision         ✅ 0.17.1>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0

**Reply by glenn-jocher** [reply]
Hello! Thank you for providing your environment details. It looks like your setup is correctly configured, which is great!

Regarding the settings reset warning, this can sometimes occur if there's an update to the `ultralytics` package or an issue with your local settings file. You can customize your settings again using the command provided, such as `yolo settings runs_dir=path/to/dir`. If you continue to experience issues with the settings, reviewing or manually editing your settings file at `/home/trupass/.config/Ultralytics/settings.yaml` might help resolve any discrepancies. For any further assistance, feel free to refer to the [Ultralytics Settings Documentation](https://docs.ultralytics.com/quickstart/#ultralytics-settings). 

Let us know if there's anything else you need! 🚀

**Comment by rlewkowicz** [comment]
I don't think you can quantize this model. Tile is not quantize friendly. I use diem, it should be the same. It's detr based. 

I think you have to pre tile the input. I'm still looking into this, but I saw this so I figured I'd comment.

**Reply by glenn-jocher** [reply]
@rlewkowicz makes an excellent point about RT-DETR models and INT8 quantization compatibility. The segmentation fault you're experiencing is likely due to the DETR architecture's complexity and certain operations that don't play well with INT8 quantization during the TFLite conversion process.

RT-DETR models use transformer-based architectures with attention mechanisms that can be challenging to quantize effectively. The segmentation fault during the INT8 calibration step suggests that some operations in the model are not compatible with the quantization process.

As an alternative, you might consider exporting to other formats that better support RT-DETR models, such as ONNX or TensorRT if you need optimized inference. You could also try exporting to regular TFLite (float32/float16) which appears to work successfully in your case.
